SessionMessagingKit/Jobs/MessageReceiveJob.swift

```diff
@@ -51,65 +51,74 @@ public enum MessageReceiveJob: JobExecutor {
+                                MessageReceiverError.duplicateControlMessage,
+                                MessageReceiverError.selfSend:
+                                break
+                                
+                            case let receiverError as MessageReceiverError where !receiverError.isRetryable:
+                                Log.error(.cat, "Permanently failed message due to error: \(error)")
+                                continue
+                                
+                            default:
+                                Log.error(.cat, "Couldn't receive message due to error: \(error)")
+                                lastError = error
+                                
+                                // We failed to process this message but it is a retryable error
+                                // so add it to the list to re-process
+                                remainingMessagesToProcess.append(messageInfo)
+                        }
+                    }
                 }
-                catch {
-                    // If the current message is a permanent failure then override it with the
-                    // new error (we want to retry if there is a single non-permanent error)
-                    switch error {
-                        // Ignore duplicate and self-send errors (these will usually be caught during
-                        // parsing but sometimes can get past and conflict at database insertion - eg.
-                        // for open group messages) we also don't bother logging as it results in
-                        // excessive logging which isn't useful)
-                        case DatabaseError.SQLITE_CONSTRAINT_UNIQUE,
-                            DatabaseError.SQLITE_CONSTRAINT,    // Sometimes thrown for UNIQUE
-                            MessageReceiverError.duplicateMessage,
-                            MessageReceiverError.duplicateControlMessage,
-                            MessageReceiverError.selfSend:
-                            break
-                        
-                        case let receiverError as MessageReceiverError where !receiverError.isRetryable:
-                            Log.error(.cat, "Permanently failed message due to error: \(error)")
-                            continue
+                
+                // If any messages failed to process then we want to update the job to only include
+                // those failed messages
+                guard !remainingMessagesToProcess.isEmpty else { return nil }
+                
+                updatedJob = try job
+                    .with(details: Details(messages: remainingMessagesToProcess))
+                    .defaulting(to: job)
+                    .upserted(db)
+                
+                return lastError
+            },
+            completion: { result in
+                // TODO: [REFACTOR] Need to test this!!!
```

Comment:
There is a `//TODO` left here, the logic looks good to me.

Reply:
Haha whoops - good catch, I've removed this (I did double check and it does work as expected)